[Merged by Bors] - feat(algebra/order/floor): lemmas about `int.floor`, `int.ceil`, `int.fract`, `round`

ocfnash opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/order/floor): lemmas about `int.floor`, `int.ceil`, `int.fract`, `round`

YaelDillies commented on src/order/basic.lean:
```diff
@@ -535,6 +535,12 @@ lemma max_rec (hx : y ≤ x → p x) (hy : x ≤ y → p y) : p (max x y) := @mi
 lemma min_rec' (p : α → Prop) (hx : p x) (hy : p y) : p (min x y) := min_rec (λ _, hx) (λ _, hy)
 lemma max_rec' (p : α → Prop) (hx : p x) (hy : p y) : p (max x y) := max_rec (λ _, hx) (λ _, hy)
 
+lemma min_def' (x y : α) : min x y = if x < y then x else y :=
+begin
+  rw [min_comm, min_def, ← ite_not],
+  simp only [not_le],
+end
```
Can we have the `max` version as well?

ocfnash replied:
Good point; done.

YaelDillies commented on src/algebra/order/floor.lean:
```diff
@@ -719,22 +753,72 @@ open int
 
-@[simp] lemma round_zero : round (0 : α) = 0 := floor_eq_iff.2 (by norm_num)
-@[simp] lemma round_one : round (1 : α) = 1 := floor_eq_iff.2 (by norm_num)
+@[simp] lemma round_one : round (1 : α) = 1 := by simp [round]
```
Can you also prove `round_nat_cast` and `round_int_cast`?

ocfnash replied:
Done.

The naming is slightly inconsistent since the corresponding lemmas for `floor` and `fract` are use `_coe` and `_coe'` suffixes. I'm not sure which we prefer.

YaelDillies commented on src/algebra/order/floor.lean:
```diff
@@ -634,6 +645,12 @@ eq_of_forall_ge_iff (λ z, by rw [neg_le, ceil_le, le_floor, int.cast_neg, neg_l
+  have : 0 < a ↔ ((0 : ℤ) : α) < a, { norm_cast, },
+  rw [this, ← lt_ceil, lt_iff_add_one_le, zero_add],
+end
+
```
Can you prove the more general lemma about `int.cast` first? Same for `floor_le_neg_one_iff`.

ocfnash replied:
I'm afraid I don't understand what you mean here. Could you spell it out?

YaelDillies replied:
```
@[simp] lemma floor_le_sub_one_iff : ⌊a⌋ - 1 ≤ n ↔ a < n := sorry
@[simp] lemma add_one_le_ceil_iff : n + 1 ≤ ⌈a⌉ ↔ n < a := sorry
```

ocfnash replied:
Thanks, now done.

YaelDillies commented on src/algebra/order/floor.lean:
```diff
@@ -531,6 +537,11 @@ lemma fract_lt_one (a : α) : fract a < 1 := sub_lt.1 $ sub_one_lt_floor _
 @[simp] lemma fract_one : fract (1 : α) = 0 :=
 by simp [fract]
 
+@[simp] lemma abs_fract : |int.fract a| = int.fract a := abs_eq_self.mpr $ fract_nonneg a
```
Can't simp do this if you tag `int.fract_nonneg`?

ocfnash replied:
Of course; thanks!

I have kept the lemma but moved the `simp` tag to `int.fract_nonneg`.

YaelDillies replied:
What about doing the same to `abs_one_sub_fract`?

ocfnash replied:
Unfortunately `simp` is not smart enough to deduce `abs_one_sub_fract` from `fract_lt_one`.

YaelDillies reviewed (commented):
Watch out for conflicts from #16025.

ocfnash committed:
Suggestions from CR

ocfnash committed:
Add lemma `int.ceil_coe'`

YaelDillies commented on src/algebra/order/floor.lean:
```diff
@@ -438,14 +444,17 @@ end
 @[simp] lemma floor_coe (z : ℤ) : ⌊(z : α)⌋ = z :=
 eq_of_forall_le_iff $ λ a, by rw [le_floor, int.cast_le]
 
+@[simp] lemma floor_coe' (n : ℕ) : ⌊(n : α)⌋ = n :=
```
Sorry, the naming convention wasn't respected originally!
```suggestion
@[simp] lemma floor_int_cast (z : ℤ) : ⌊(z : α)⌋ = z :=
eq_of_forall_le_iff $ λ a, by rw [le_floor, int.cast_le]

@[simp] lemma floor_nat_cast (n : ℕ) : ⌊(n : α)⌋ = n :=
```

ocfnash committed:
CR suggestion: add lemmas `int.floor_le_sub_one_iff`, `int.add_one_le_ceil_iff`

ocfnash committed:
Update src/algebra/order/floor.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

ocfnash committed:
Update naming: change `_coe`, `_coe'` suffixes to `_int_cast`, `_nat_cast`.

ocfnash committed:
Update naming 2: change `_coe`, `_coe'` suffixes to `_int_cast`, `_nat_cast`.

The sequel

ocfnash committed:
Fix long lines

ocfnash committed:
Merge branch 'master' into ocfnash/floor_rings

jcommelin reviewed (approved):
Thanks :tada:

bors merge

ocfnash committed:
Fix after merging master

The relevant changes arrived in master via https://github.com/leanprover-community/mathlib/pull/16025

ocfnash committed:
Merge branch 'master' into ocfnash/floor_rings
